File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.py
import re
from collections import defaultdict
from itertools import combinations

def solution(orders, course):
    answer = []
    size = len(orders)
    
    orders.sort(key=lambda x:len(x))   #메뉴 길이가 짧은 순서대로 정렬
    
    orders_set = []  #각 메뉴별 집합 리스트
        
    for i in range(size):
        orders_set.append(set(orders[i]))

    result = {}     #음식개수별로 counter리스트를 만듦
    for c in course:
        result[c] = defaultdict(int)
        
    for i in range(size - 1):
        for j in range(i + 1, size):
            menu =  compare_string(orders_set[i], orders_set[j])  #두 사람의 겹치는 코스
            if menu == '' or len(menu) < 2:  #메뉴가 2개 이상이어야 함
                continue
            
            for count in course:
                for m in combinations(menu, count):   #해당 메뉴로 가능한 모든 음식 조합 개수 세기(course 리스트의 값들 만큼)
                    menu_comb = "".join(sorted(m))
                    result[len(menu_comb)][menu_comb] += 1  #result에서 해당하는 메뉴개수의 counter배열에 음식 개수 + 1 업데이트
        
    for menu_count in result:  #모든 조합결과를 비교해봄
        if len(result[menu_count]) == 0:
            continue
            
        max_count = max(result[menu_count].values())  #해당 메뉴개수의 dict에서 가장 큰 값
        
        for menu, count in result[menu_count].items():
            if count == max_count:  #가장 많이 주문한 메뉴에 해당하면 최종주문목록에 추가함
                answer.append(menu)
    
    return sorted(answer)

def compare_string(first, second):
    result = ''
    
    for i in first:
        if i in second:
            result += i

    return "".join(sorted(result))

File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.java
import java.util.*;

class Solution {
    private static Map<String, Integer> orderList;
    private static Set<Integer> courseCount;
    private static Map<Integer, Integer> courseIndex;
    
    public String[] solution(String[] orders, int[] course) {
        orderList = new HashMap<>();
        courseCount = new HashSet<>();
        courseIndex = new HashMap<>();

        for (int i = 0; i < course.length; i++) {
            courseCount.add(course[i]);
            courseIndex.put(course[i], i);
        }
        
        for (int i = 0; i < orders.length; i++) {
            char[] newOrder = orders[i].toCharArray();
            Arrays.sort(newOrder);
            orders[i] = new String(newOrder);
        }
        
        int[] maxCount = new int[course.length];
        
        for (String order: orders) {
            dfs("", order, 0);
        }
        
        for (String key: orderList.keySet()) {
            if (orderList.get(key) >= 2) {
                int index = courseIndex.get(key.length());
                maxCount[index] = Math.max(maxCount[index], orderList.get(key));
            }
        }

        List<String> result = new ArrayList<>();
        
        for (String key: orderList.keySet()) {
            int index = courseIndex.get(key.length());
            if (orderList.get(key) == maxCount[index]) {
                result.add(key);
            }
        }
        
        String[] answer = new String[result.size()];
        for (int i = 0; i < result.size(); i++) {
            answer[i] = result.get(i);
        }
        
        Arrays.sort(answer, (o1, o2) -> o1.compareTo(o2));
        
        return answer;
    }
    
    private static void dfs(String result, String order, int current) {
        if (courseCount.contains(result.length())) {
            int count = 1;
            if (orderList.containsKey(result)) {
                count = orderList.get(result) + 1;
            }
            orderList.put(result, count);
        }
        
        if (current == order.length()) return;
        
        for (int i = current; i < order.length(); i++) {
            dfs(result + order.charAt(i), order, i + 1);
        }
    }
}